### PR TITLE
feat: implement floor gas + calldata gas and improve gas calculation

### PR DESCRIPTION
**Pull request by 0xpolarzero**
## Description
Implements EIP-7623 (Floor Gas) and refactors gas calculation logic for transactions. This PR adds support for calculating the minimum gas a transaction must consume based on calldata tokens, which is required for the Prague hardfork. The implementation includes:

1. Adding EIP-7623 to the Prague hardfork EIP list
2. Implementing token-based calldata gas calculation
3. Adding floor gas cost calculation for Prague
4. Refactoring intrinsic gas calculation to be more modular
5. Ensuring transactions consume at least the floor gas amount

The PR also adds comprehensive tests for all the new gas calculation functions, including edge cases and regression tests.

## Type of Change
- [x] 🎉 New feature (non-breaking change which adds functionality)
- [ ] ✅ Test additions or updates

## Testing
- [x] `zig build test` passes
- [x] `zig build` completes successfully
- [x] All existing tests pass
- [x] Added new tests for changes (if applicable)

## Checklist
- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by 0xpolarzero** [comment]
* **#825** <a href="https://app.graphite.dev/github/pr/evmts/guillotine/825?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/evmts/guillotine/825?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/evm.zig [hunk 1]
```diff
@@ -502,13 +509,21 @@ pub fn Evm(config: EvmConfig) type {
...
                 self.gas_refund_counter = 0;
             }
 
+            var gas_consumed = gas_limit - result.gas_left;
+            // EIP-7623 (Prague): Ensure at least floor_gas_cost is consumed
+            if (gas_consumed < floor_gas_cost) {
+                gas_consumed = floor_gas_cost;
+                if (!comptime config.disable_gas_checks) {
+                    result.gas_left = gas_limit - floor_gas_cost;
+                }
```
Comment (roninjin10) [inline src/evm.zig:523]: this doesn't look right to me

File: src/evm.zig [hunk 2]
```diff
@@ -502,13 +509,21 @@ pub fn Evm(config: EvmConfig) type {
...
             if (result.success) {
-                result.gas_left = config.eips.eip_3529_apply_gas_refund(initial_gas, result.gas_left, self.gas_refund_counter);
+                result.gas_left = config.eips.eip_3529_apply_gas_refund(gas_limit, result.gas_left, self.gas_refund_counter);
                 // Reset refund counter for next transaction
                 self.gas_refund_counter = 0;
             }
 
+            var gas_consumed = gas_limit - result.gas_left;
+            // EIP-7623 (Prague): Ensure at least floor_gas_cost is consumed
+            if (gas_consumed < floor_gas_cost) {
```
Comment (roninjin10) [inline src/evm.zig:519]: nit using std.math.max would be more readable

File: src/evm.zig [hunk 3]
```diff
@@ -448,25 +449,31 @@ pub fn Evm(config: EvmConfig) type {
...
+            // gas += authorization_list_num * PER_EMPTY_ACCOUNT_COST (25000);
+            const calldata_tokens = config.eips.tokens_in_calldata(params.getInput());
+            const intrinsic_gas_cost = config.eips.intrinsic_gas_cost(params.isCreate());
+            const calldata_gas_cost = config.eips.calldata_gas_cost(calldata_tokens, params.isCreate(), params.getInput().len);
+            const floor_gas_cost = config.eips.floor_gas_cost(calldata_tokens);
+
+            const initial_gas_cost = intrinsic_gas_cost + calldata_gas_cost;
+            if (comptime !config.disable_gas_checks) {
+                if (gas_limit < initial_gas_cost) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
+                if (gas_limit < floor_gas_cost) return CallResult.failure(self.getCallArenaAllocator(), 0) catch unreachable;
```
Comment (roninjin10) [inline src/evm.zig:474]: This doesn't look right to me we still need to deal with not underflowing. Might be a good idea to remove disable gas cost as an option right now seems to be randomly implemented without much thought and too much vibes everywhere

**Review by roninjin10 — CHANGES_REQUESTED** [review]
Is the places where it looks like a bug to me a bug? What test is this pr making pass or should this be adding a test?
